Index: Makefile.am
===================================================================
--- Makefile.am	(revision 16470)
+++ Makefile.am	(working copy)
@@ -10,7 +10,8 @@
 		lackey \
 		none \
 		helgrind \
-		drd
+		drd \
+		verrou
 
 EXP_TOOLS = 	exp-sgcheck \
 		exp-bbv \
Index: README
===================================================================
--- README	(revision 16470)
+++ README	(working copy)
@@ -1,4 +1,26 @@
 
+Specific instructions for Verrou
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You are currently looking at a version of valgrind which has been patched to
+contain the verrou tool. Please look at specific instructions contained in the file:
+
+    verrou/README.md
+
+or the verrou github page:
+
+    https://github.com/edf-hpc/verrou
+
+
+For the very impatient, you should probably run the following commands:
+
+    ./autogen.sh
+    ./configure --enable-only64bit --enable-verrou-fma --prefix=PREFIX
+    make
+    make install
+    source PREFIX/env.sh
+
+
 Release notes for Valgrind
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 If you are building a binary package of Valgrind for distribution,
Index: configure.ac
===================================================================
--- configure.ac	(revision 16470)
+++ configure.ac	(working copy)
@@ -8,7 +8,7 @@
 ##------------------------------------------------------------##
 
 # Process this file with autoconf to produce a configure script.
-AC_INIT([Valgrind],[3.13.0],[valgrind-users@lists.sourceforge.net])
+AC_INIT([Valgrind],[3.13.0+verrou-dev],[valgrind-users@lists.sourceforge.net])
 AC_CONFIG_SRCDIR(coregrind/m_main.c)
 AC_CONFIG_HEADERS([config.h])
 AM_INIT_AUTOMAKE([foreign subdir-objects])
@@ -4353,6 +4353,10 @@
                [test x$ac_have_shared_pointer_annotation = xyes])
 
 
+# Verrou
+m4_include(verrou/configure.ac)
+
+
 #----------------------------------------------------------------------------
 # Ok.  We're done checking.
 #----------------------------------------------------------------------------
@@ -4445,6 +4449,8 @@
    exp-bbv/tests/arm-linux/Makefile
    exp-dhat/Makefile
    exp-dhat/tests/Makefile
+   verrou/Makefile
+   verrou/tests/Makefile
    shared/Makefile
    solaris/Makefile
 ])
Index: docs/lib/vg_basic.css
===================================================================
--- docs/lib/vg_basic.css	(revision 16470)
+++ docs/lib/vg_basic.css	(working copy)
@@ -65,3 +65,39 @@
  border:      solid 1px #ffde84; 
 }
 
+div.note,
+div.warning {
+    margin-left: 0 !important;
+    margin-right: 0 !important;
+    border-radius: 4px;
+    padding: 1em;
+}
+
+div.note h3,
+div.warning h3 {
+    font-size: 100%;
+    font-weight: bold;
+    margin-top: 0;
+}
+
+div.note p,
+div.warning p {
+    margin: 0;
+}
+
+div.note {
+    background-color: #d9edf7;
+    border: 1px solid #bce8f1;
+}
+div.note, div.note h3 {
+    color: #31708f;
+}
+
+div.warning {
+    background-color: #fcf8e3;
+    border: 1px solid #faebcc;
+}
+
+div.warning, div.warning h3 {
+    color: #8a6d3b;
+}
Index: docs/xml/manpages-index.xml
===================================================================
--- docs/xml/manpages-index.xml	(revision 16470)
+++ docs/xml/manpages-index.xml	(working copy)
@@ -43,5 +43,9 @@
 <xi:include href="../../coregrind/docs/vgdb-manpage.xml"
     parse="xml" xmlns:xi="http://www.w3.org/2001/XInclude" />
 
+<!-- verrou_dd -->
+<xi:include href="../../verrou/docs/verrou_dd-manpage.xml"
+    parse="xml" xmlns:xi="http://www.w3.org/2001/XInclude" />
+
 </set>
 
Index: docs/xml/manual.xml
===================================================================
--- docs/xml/manual.xml	(revision 16470)
+++ docs/xml/manual.xml	(working copy)
@@ -44,6 +44,8 @@
       xmlns:xi="http://www.w3.org/2001/XInclude" />      
   <xi:include href="../../lackey/docs/lk-manual.xml" parse="xml"  
       xmlns:xi="http://www.w3.org/2001/XInclude" />
+  <xi:include href="../../verrou/docs/vr-manual.xml" parse="xml"
+      xmlns:xi="http://www.w3.org/2001/XInclude" />
   <xi:include href="../../none/docs/nl-manual.xml" parse="xml"  
       xmlns:xi="http://www.w3.org/2001/XInclude" />
 
Index: docs/xml/valgrind-manpage.xml
===================================================================
--- docs/xml/valgrind-manpage.xml	(revision 16470)
+++ docs/xml/valgrind-manpage.xml	(working copy)
@@ -230,6 +230,8 @@
 
 </refsect1>
 
+<xi:include href="../../verrou/docs/valgrind-manpage.xml"
+            xmlns:xi="http://www.w3.org/2001/XInclude" />
 
 <refsect1 id="see_also">
 <title>See Also</title>
Index: docs/xml/vg-entities.xml
===================================================================
--- docs/xml/vg-entities.xml	(revision 16470)
+++ docs/xml/vg-entities.xml	(working copy)
@@ -6,7 +6,7 @@
 
 <!-- valgrind release + version stuff -->
 <!ENTITY rel-type    "Release">
-<!ENTITY rel-version "3.13.0">
+<!ENTITY rel-version "3.13.0+verrou-dev">
 <!ENTITY rel-date    "15 June 2017">
 
 <!-- where the docs are installed -->
