diff --git a/Makefile.am b/Makefile.am
index 08db83401..cfcd7679f 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -14,6 +14,8 @@ TOOLS = \
 		lackey \
 		none
 
+TOOLS += verrou
+
 EXP_TOOLS = \
 		exp-bbv
 
diff --git a/README b/README
index 511027ae9..0f32e73ea 100644
--- a/README
+++ b/README
@@ -1,4 +1,26 @@
 
+Specific instructions for Verrou
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You are currently looking at a version of valgrind which has been patched to
+contain the verrou tool. Please look at specific instructions contained in the file:
+
+    verrou/README.md
+
+or the verrou github page:
+
+    https://github.com/edf-hpc/verrou
+
+
+For the very impatient, you should probably run the following commands:
+
+    ./autogen.sh
+    ./configure --enable-only64bit --enable-verrou-fma --prefix=PREFIX
+    make
+    make install
+    source PREFIX/env.sh
+
+
 Release notes for Valgrind
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 If you are building a binary package of Valgrind for distribution,
diff --git a/configure.ac b/configure.ac
index b86e80220..f67bafdf1 100755
--- a/configure.ac
+++ b/configure.ac
@@ -8,7 +8,7 @@
 ##------------------------------------------------------------##
 
 # Process this file with autoconf to produce a configure script.
-AC_INIT([Valgrind],[3.17.0],[valgrind-users@lists.sourceforge.net])
+AC_INIT([Valgrind],[3.17.0+verrou-dev],[valgrind-users@lists.sourceforge.net])
 AC_CONFIG_SRCDIR(coregrind/m_main.c)
 AC_CONFIG_HEADERS([config.h])
 AM_INIT_AUTOMAKE([foreign dist-bzip2 subdir-objects])
@@ -4898,6 +4898,10 @@ AM_CONDITIONAL([HAVE_SHARED_POINTER_ANNOTATION],
                [test x$ac_have_shared_pointer_annotation = xyes])
 
 
+# Verrou
+m4_include(verrou/configure.ac)
+
+
 #----------------------------------------------------------------------------
 # Ok.  We're done checking.
 #----------------------------------------------------------------------------
@@ -4920,6 +4924,8 @@ AC_CONFIG_FILES([
    auxprogs/Makefile
    mpi/Makefile
    coregrind/Makefile 
+   verrou/Makefile
+   verrou/tests/Makefile
    memcheck/Makefile
    memcheck/tests/Makefile
    memcheck/tests/common/Makefile
diff --git a/docs/lib/vg_basic.css b/docs/lib/vg_basic.css
index 49367fed2..6dc3e5f7e 100644
--- a/docs/lib/vg_basic.css
+++ b/docs/lib/vg_basic.css
@@ -65,3 +65,39 @@ blockquote {
  border:      solid 1px #ffde84; 
 }
 
+div.note,
+div.warning {
+    margin-left: 0 !important;
+    margin-right: 0 !important;
+    border-radius: 4px;
+    padding: 1em;
+}
+
+div.note h3,
+div.warning h3 {
+    font-size: 100%;
+    font-weight: bold;
+    margin-top: 0;
+}
+
+div.note p,
+div.warning p {
+    margin: 0;
+}
+
+div.note {
+    background-color: #d9edf7;
+    border: 1px solid #bce8f1;
+}
+div.note, div.note h3 {
+    color: #31708f;
+}
+
+div.warning {
+    background-color: #fcf8e3;
+    border: 1px solid #faebcc;
+}
+
+div.warning, div.warning h3 {
+    color: #8a6d3b;
+}
diff --git a/docs/xml/manpages-index.xml b/docs/xml/manpages-index.xml
index 17a81c99f..2086f49eb 100644
--- a/docs/xml/manpages-index.xml
+++ b/docs/xml/manpages-index.xml
@@ -78,6 +78,13 @@
     parse="xml" xmlns:xi="http://www.w3.org/2001/XInclude" />
 </chapter>
 
+<!-- verrou_dd -->
+<chapter>
+<title>verrou_dd</title>
+<xi:include href="../../verrou/docs/verrou_dd-manpage.xml"
+    parse="xml" xmlns:xi="http://www.w3.org/2001/XInclude" />
+</chapter>
+
 </book>
 </set>
 
diff --git a/docs/xml/manual.xml b/docs/xml/manual.xml
index 518be2964..5929b16a5 100644
--- a/docs/xml/manual.xml
+++ b/docs/xml/manual.xml
@@ -40,6 +40,8 @@
       xmlns:xi="http://www.w3.org/2001/XInclude" />
   <xi:include href="../../lackey/docs/lk-manual.xml" parse="xml"  
       xmlns:xi="http://www.w3.org/2001/XInclude" />
+  <xi:include href="../../verrou/docs/vr-manual.xml" parse="xml"
+      xmlns:xi="http://www.w3.org/2001/XInclude" />
   <xi:include href="../../none/docs/nl-manual.xml" parse="xml"  
       xmlns:xi="http://www.w3.org/2001/XInclude" />
   <xi:include href="../../exp-bbv/docs/bbv-manual.xml" parse="xml"  
diff --git a/docs/xml/valgrind-manpage.xml b/docs/xml/valgrind-manpage.xml
index 3c893c680..d36eadec3 100644
--- a/docs/xml/valgrind-manpage.xml
+++ b/docs/xml/valgrind-manpage.xml
@@ -235,6 +235,9 @@
             xmlns:xi="http://www.w3.org/2001/XInclude" />
 </refsect1>
 
+<xi:include href="../../verrou/docs/valgrind-manpage.xml"
+            xmlns:xi="http://www.w3.org/2001/XInclude" />
+
 <refsect1 id="see_also">
 <title>See Also</title>
 

diff --git a/docs/xml/vg-entities.xml b/docs/xml/vg-entities.xml
index 7d04ee795..a105f0b50 100644
--- a/docs/xml/vg-entities.xml
+++ b/docs/xml/vg-entities.xml
@@ -6,7 +6,7 @@
 
 <!-- valgrind release + version stuff -->
 <!ENTITY rel-type    "Release">
-<!ENTITY rel-version "3.17.0">
+<!ENTITY rel-version "3.17.0+verrou-dev">
 <!ENTITY rel-date    "19 Mar 2021">
 
 <!-- where the docs are installed -->
